lib/charms/opensearch/v0/opensearch_backups.py

```diff
@@ -0,0 +1,641 @@
+            else:
+                self.charm.unit.status = BlockedStatus(
+                    "Unexpected error during plugin configuration, check the logs"
+                )
+                # There was an unexpected error, log it and block the unit
+                logger.error(e)
+            event.defer()
+        self.charm.unit.status = ActiveStatus()
+
+    def apply_post_restart_if_needed(self) -> None:
+        """Runs the post restart routine and API calls needed to setup/disable backup.
+
+        This method should be called by the charm in its restart callback resolution.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if self._check_missing_s3_config_completeness():
+            # Always check if a relation actually exists and if options are available
+            # in this case, seems one of the conditions above is not yet present
+            # abandon this restart event, as it will be called later once s3 configuration
+            # is correctly set
+            return
+
+        if self.charm.plugin_manager.status(self._plugin) not in [
+            PluginState.ENABLED,
+            PluginState.WAITING_FOR_UPGRADE,
+        ]:
+            # Configuration is available, but this restart was not done to implement the
+            # new backup configs. Abandon the event.
+            return
+
+        if self._service_already_registered():
+            # Finally, check if all the options above are true AND the configuration has
+            # already been applied. If that is the case, then also leave the event.
+            return
+
+        # Backup relation has been recently made available with all the parameters needed.
+        # Steps:
+        #     (1) set up as maintenance;
+        self.charm.unit.status = MaintenanceStatus("Configuring backup service...")
+        #     (2) run the request; and
+        state = self._register_snapshot_repo()
+        #     (3) based on the response, set the message status
+        if state == BackupServiceState.SUCCESS:
+            self.charm.unit.status = ActiveStatus("Backup is active")
+        else:
+            self.charm.unit.status = BlockedStatus(f"Backup setup failed with {state}")
+
+    def on_s3_depart(self, event: EventBase) -> None:
+        """Processes the departure of s3.
+
+        It runs the reverse process of on_s3_change:
+        1) Check if the cluster is currently taking a snapshot, if yes, set status as blocked
+           and defer this event.
+        2) If leader, run API calls to signal disable is needed
+        3) Update opensearch.yml, CA truststore and keystore
+        4) Emit a restart event
+        """
+        self.charm.unit.status = MaintenanceStatus("Disabling backup service...")
+        snapshot_status = self._check_snapshot_status()
+        if snapshot_status in [
+            BackupServiceState.SNAPSHOT_IN_PROGRESS,
+            BackupServiceState.SNAPSHOT_PARTIALLY_TAKEN,
```

Comment:
From the docs:
```
PARTIAL: At least one shard failed to store successfully. Can only occur if you set partial to true when taking the snapshot.
```

So, it may happen, the question is what should we do. I am removing it from this particular check, as we should be able to remove a relation correctly if we have a partially taken snapshot. I will add a warn log to raise the remove is happening but we had a partially taken snapshot.